File: wrapper/ios/ext/Feature/ObjectDetection/YOLOv8/YOLOv8.swift
import Foundation

public class YOLOv8: PipelineFeature {
    
    private let wrapper: YOLOv8Wrapper
    private let model: PipelineInferenceModel
    
    public init(cocoYamlFilePath: String? = nil,
         model: PipelineInferenceModel? = nil) {
        
        let yamlURL = Bundle.main.url(forResource: "coco", withExtension: "yaml")!
        self.wrapper = YOLOv8Wrapper(yamlURL.absoluteString)
        
        self.model = model ?? ZeticMLangeModelWrapper(
            "ztp_97aT0F0HtHQ5Q3dasRCIAoxKH0O0YKJUyvOB",
            "b9f5d74e6f644288a32c50174ded828e"
        )
    }
    
    public func preprocess(_ input: CameraFrame) -> [Data] {
        return input.withUnsafeBytes { frameBytes in
            return [wrapper.featurePreprocess(
                frameBytes.baseAddress!,
                input.width,
                input.height,
                input.bytesPerRow
            )]
        }
    }
    
    public func postprocess(_ output: [Data]) -> ObjectDetectionResult {
        return output[0].withUnsafeBytes { (buffer: UnsafeRawBufferPointer) in
            let pointer = buffer.bindMemory(to: UInt8.self).baseAddress
            return if let uint8Pointer = pointer {
                ObjectDetectionResult(wrapper.featurePostprocess(uint8Pointer))
            } else {
                ObjectDetectionResult([])
            }
        }
    }
    
    public func run(_ input: CameraFrame) -> ObjectDetectionResult {
        let preprocessed = preprocess(input)
        let output = model.inference(preprocessed)
        return postprocess(output)
    }
}

File: wrapper/android/ext/src/main/java/com/zeticai/mlange/feature/objectdetection/yolov8/YOLOv8.kt
package com.zeticai.mlange.feature.objectdetection.yolov8

import android.content.Context
import com.zeticai.mlange.core.common.DataUtils
import com.zeticai.mlange.core.tensor.Tensor
import com.zeticai.mlange.feature.PipelineFeature
import com.zeticai.mlange.feature.PipelineInferenceModel
import com.zeticai.mlange.feature.ZeticMLangeModelWrapper
import com.zeticai.mlange.feature.objectdetection.ObjectDetectionResult
import com.zeticai.mlange.feature.vision.OpenCVImageUtilsWrapper
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer

class YOLOv8(
    private val context: Context,
    cocoYamlFilePath: String = findCocoYamlFilePath(context),
    private val model: PipelineInferenceModel = ZeticMLangeModelWrapper(
        context,
        "ztp_97aT0F0HtHQ5Q3dasRCIAoxKH0O0YKJUyvOB",
        "b9f5d74e6f644288a32c50174ded828e"
    ),
) : PipelineFeature<ObjectDetectionResult> {
    val wrapper = YOLOv8Wrapper(cocoYamlFilePath)
    private val openCVImageUtilsWrapper = OpenCVImageUtilsWrapper()

    override fun preprocess(input: Array<ByteArray>): Array<ByteArray> {
        val imagePtr = openCVImageUtilsWrapper.frame(input[0], 90)
        return arrayOf(wrapper.preprocess(imagePtr))
    }

    override fun postprocess(output: Array<ByteBuffer>): ObjectDetectionResult {
        return ObjectDetectionResult(wrapper.postprocess(output[0]).value)
    }

    override fun run(input: Array<ByteArray>): ObjectDetectionResult {
        val pre = preprocess(input)
        val output = model.inference(pre.map {
            Tensor.of(it)
        }.toTypedArray())
        return postprocess(output.map {
            it.data
        }.toTypedArray())
    }

    companion object {
        private fun findCocoYamlFilePath(context: Context): String {
            val cocoYamlSamplePath = "coco.yaml"
            copyFileFromAssetsToData(context, cocoYamlSamplePath)
            val cocoYamlFile = File(context.filesDir, cocoYamlSamplePath)
            return cocoYamlFile.absolutePath
        }

        private fun copyFileFromAssetsToData(context: Context, fileName: String) {
            val assetManager = context.assets
            val file = assetManager.open(fileName)
            val outFile = File(context.filesDir, fileName)
            val out = FileOutputStream(outFile)

            val buffer = ByteArray(1024)
            var read: Int
            while ((file.read(buffer).also { read = it }) != -1) {
                out.write(buffer, 0, read)
            }

            file.close()
            out.flush()
            out.close()
        }
    }
}
